proposals/explicit-api-mode.md

```diff
@@ -0,0 +1,108 @@
+# Explicit API mode
+
+* **Type**: Design proposal
+* **Author**: Ilya Gorbunov
+* **Contributors**: Leonid Startsev, Roman Elizarov
+* **Status**: Prototype
+* **Original proposal and discussion**: [KEEP-45](https://github.com/Kotlin/KEEP/issues/45)
+
+## Synopsis
+
+Provide an 'Explicit API' mode in the compiler which helps library authoring.
+Such mode should prevent delivery of unintended pubic API/ABI to the clients by requiring explicit visibility and return types on declarations.
+
+## Motivation
+
+There were a couple of hot discussions about the default visibility before release:
+
+* https://discuss.kotlinlang.org/t/public-by-default-for-classes/110
+* https://discuss.kotlinlang.org/t/kotlins-default-visibility-should-be-internal/1400
+
+While such a decision is convenient for application programming, the main concern against public-by-default visibility was that it becomes too easy for library authors to expose something accidentally, release it, and then have to make breaking changes to hide it back.
+
+## Proposal
+
+Introduce 'Explicit API' compiler mode. Compilation in such mode differs from the default mode in the following aspects:
+
+* Compiler requires you to specify explicit visibility for a declaration when leaving default visibility would result in exposing that declaration to the public API surface.
+
+* Compiler requires you to specify the explicit type of property/function when it is exposed to the public API surface.
+
+* Compiler warns you when exposed to public API surface declaration does not have a KDoc.
+
+See the definition of public API [here](https://github.com/JetBrains/kotlin/blob/master/libraries/tools/binary-compatibility-validator/ReadMe.md#what-constitutes-the-public-api).
```

Comment:
I'd suggest to give more details here. In particular I would clearly distinguish "Public API" vs "Published API" where "Published API" should be defined as "Public API" + "internal @PublishedApi" declarations. Why? Because "Published API" itself should not require KDocs. Only when its part of "Public API" then should require KDocs. 

Also, as a stylistic note the world "surface" looks redundant in text. I suggest to systematically replace phrases like "public API surface" with a simpler "public API"